WooCommerce/Classes/ViewModels/OrderDetailsViewModel/OrderDetailsViewModel.swift

```diff
@@ -0,0 +1,445 @@
+            displayEmailComposerIfPossible(from: viewController)
+        case .billingPhone:
+            displayContactCustomerAlert(from: viewController)
+        default:
+            break
+        }
+    }
+}
+
+
+// MARK: - MFMailComposeViewControllerDelegate Conformance
+//
+private extension OrderDetailsViewModel {
+    func displayEmailComposerIfPossible(from: UIViewController) {
+        emailComposer.displayEmailComposerIfPossible(for: order, from: from)
+    }
+
+    /// Displays an alert that offers several contact methods to reach the customer: [Phone / Message]
+    ///
+    func displayContactCustomerAlert(from: UIViewController) {
+        guard let sourceView = from.view else {
+            return
+        }
+
+        let actionSheet = UIAlertController(title: nil, message: nil, preferredStyle: .actionSheet)
+        actionSheet.view.tintColor = StyleManager.wooCommerceBrandColor
+
+        actionSheet.addCancelActionWithTitle(ContactAction.dismiss)
+        actionSheet.addDefaultActionWithTitle(ContactAction.call) { [weak self] _ in
+            guard let self = self else {
+                return
+            }
+
+            guard let phoneURL = self.order.billingAddress?.cleanedPhoneNumberAsActionableURL else {
+                return
+            }
+
+            WooAnalytics.shared.track(.orderDetailCustomerPhoneOptionTapped)
+            self.callCustomerIfPossible(at: phoneURL)
+        }
+
+        actionSheet.addDefaultActionWithTitle(ContactAction.message) { [weak self] _ in
+            WooAnalytics.shared.track(.orderDetailCustomerSMSOptionTapped)
+            self?.displayMessageComposerIfPossible(from: from)
+        }
+
+        let popoverController = actionSheet.popoverPresentationController
+        popoverController?.sourceView = sourceView
+        popoverController?.sourceRect = sourceView.bounds
+
+        from.present(actionSheet, animated: true)
+
+        WooAnalytics.shared.track(.orderDetailCustomerPhoneMenuTapped)
+    }
+
+    /// Attempts to perform a phone call at the specified URL
+    ///
+    func callCustomerIfPossible(at phoneURL: URL) {
+        guard UIApplication.shared.canOpenURL(phoneURL) else {
+            return
+        }
+
+        UIApplication.shared.open(phoneURL, options: [:], completionHandler: nil)
+        WooAnalytics.shared.track(.orderContactAction, withProperties: ["id": order.orderID,
+                                                                        "status": self.order.statusKey,
+                                                                        "type": "call"])
+
+    }
+}
+
+private extension OrderDetailsViewModel {
+    func displayMessageComposerIfPossible(from: UIViewController) {
+        messageComposer.displayMessageComposerIfPossible(order: order, from: from)
+    }
+}
+
+
+// MARK: - Syning data. Yosemite related stuff
+extension OrderDetailsViewModel {
+    func syncOrder(onCompletion: ((Order?, Error?) -> ())? = nil) {
```

Comment:
It's interesting that the data syncing is happening in the ViewModel. You talked about maybe adding a presenter once and now I'm starting to see why we'd need one (or something similar to it, broken out into a separate file). It's not really a ViewModel's responsibility to sync data.

Reply:
Exactly. In this case, I am using the ViewModel suffix with a lot of latitude. This class is doing more than the traditional role of a view model, but I decided to stop with the refactoring here, to keep it manageable. Eventually, we'd see the need to extract the syncing, but at least I feel like now we are in a better position to do so.

Reply:
>I decided to stop with the refactoring here, to keep it manageable. 
yep :) I appreciate it too! This code is teaching me so many things 😁